{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "scientific-chuck",
   "metadata": {},
   "source": [
    "# Ets in putative heart enhancers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "appreciated-content",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "#########################################################################################\n",
    "# These methods can be found on my gitub\n",
    "# see https://github.com/jsolvason/js/tree/6623ca2c3ee871eca8defbac5b85fcb4036eb7a4\n",
    "#    for methods used here at time of publication\n",
    "\n",
    "import jsDna as jsd\n",
    "import js"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "proud-stability",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################################################\n",
    "# Download data form wei 2010, \n",
    "# http://the_brain.bwh.harvard.edu/uniprobe/detailsDef.php?id=414\n",
    "# Mouse ETS1 pbm data\n",
    "def loadEts():\n",
    "    ref  =f'/projects/ps-farleylab/users/solvason/ref/binding-affinity/ets/parsed_Ets1_8mers.txt'              \n",
    "    Seq2EtsAff  = {line.split('\\t')[0]:float(line.split('\\t')[1]) for line in open(ref,'r').readlines()}\n",
    "    return Seq2EtsAff\n",
    "\n",
    "Ets2Aff=loadEts()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "featured-multimedia",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################################################\n",
    "# Load hg38  genome\n",
    "\n",
    "d=\"/projects/ps-farleylab/users/solvason/ref/genome/\"\n",
    "def loadHg38():\n",
    "    file_genome=f'{d}/human/hg38/hg38.fa'\n",
    "    Handle = open(file_genome, \"r\")\n",
    "    chr2seq = SeqIO.to_dict(SeqIO.parse(Handle, \"fasta\"))\n",
    "    chr2seq = {chrom:str(chr2seq[chrom].seq).upper() for chrom in chr2seq}\n",
    "    return chr2seq\n",
    "\n",
    "# Load 2019 ciona genome\n",
    "def loadCi19():\n",
    "    file_genome=f'{d}/ciona/2019/HT.Ref.fasta'\n",
    "    Handle = open(file_genome, \"r\")\n",
    "    chr2seq = SeqIO.to_dict(SeqIO.parse(Handle, \"fasta\"))\n",
    "    chr2seq = {chrom:str(chr2seq[chrom].seq).upper() for chrom in chr2seq}\n",
    "    return chr2seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "enormous-polls",
   "metadata": {},
   "source": [
    "# Load genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "isolated-windows",
   "metadata": {},
   "outputs": [],
   "source": [
    "G2Chr2Seq={}\n",
    "\n",
    "G2Chr2Seq['Ci19']=loadCi19()\n",
    "G2Chr2Seq['hg38']=loadHg38()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wooden-sentence",
   "metadata": {},
   "source": [
    "# Analyze ets affinities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "liable-mandate",
   "metadata": {},
   "outputs": [],
   "source": [
    "Ets2RoundedAff={kmer:round(float(aff),2) for kmer,aff in Ets2Aff.items()}\n",
    "js.dprint(Ets2RoundedAff,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "northern-smith",
   "metadata": {},
   "outputs": [],
   "source": [
    "Genome2Organism={\n",
    "    'Ci19':'Ciona',\n",
    "    'hg38':'Human'\n",
    "}\n",
    "\n",
    "d='/home/solvason/projects/ets/heart/0-affinity-optimization-potential/'\n",
    "Genome2Name2Bed={\n",
    "    'Ci19':{'Heart-Open-Chromatin-Racioppi'   :f'{d}/ciona/racciopi-atac-mespPositive/LacZ_6hpf_13_peaks_noexon_top90.bed'},\n",
    "    \n",
    "    'hg38': {\n",
    "        'Heart-Put-Dev-En-Benaglio-Dantonio-May-Richter':'/home/solvason/projects/ets/heart/0-affinity-optimization-potential/human/7-heart-fetal-enhancers-multipleDatasets-forcellpaper/merged.bed.sorted',\n",
    "    }\n",
    "}\n",
    "\n",
    "Name2Xlabel={\n",
    "    'Heart-Put-Dev-En-Benaglio-Dantonio-May-Richter':'Heart-Put-Dev-En\\nBenaglio-Dantonio\\nMay-Richter',\n",
    "    'Heart-Open-Chromatin-Racioppi':'Heart-Open-Chromatin\\nRacioppi',\n",
    "}\n",
    "\n",
    "scn5a='GCTTCCGG' \n",
    "mars2='AAGGAAAT'\n",
    "ryr2='TCTTCCGG'\n",
    "zfhx3='ACGGAAAC'\n",
    "sh3yl1='CAGGAAGC'\n",
    "diseaseKmers =[scn5a,mars2,ryr2,zfhx3,sh3yl1]\n",
    "xAdjustCustom=[0,+1,-1,-1,+1]# see cell below to adjust x axis so the dots dont overlap on scatter\n",
    "\n",
    "Genome2Name2DiseaseKmers={\n",
    "    'Ci19':{'Heart-Open-Chromatin-Racioppi' : ([],[])},\n",
    "    'hg38':{\n",
    "        'Heart-Put-Dev-En-Benaglio-Dantonio-May-Richter':                    (diseaseKmers,xAdjustCustom),\n",
    "        'Heart-Put-Dev-En-Benaglio-Dantonio-May-Richter-associated-GWAS':     (diseaseKmers,xAdjustCustom),\n",
    "    },\n",
    "    \n",
    "}\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "handled-radical",
   "metadata": {},
   "outputs": [],
   "source": [
    "etsTfCores=set(['GGAA','GGAT','TTCC','ATCC'])\n",
    "\n",
    "\n",
    "def checkIfSnvCausesDramaticIncrease(ref8mer,Ets2RoundedAff,foldChangeCutoff):\n",
    "    \n",
    "    kmer=ref8mer\n",
    "    aff=Ets2RoundedAff[kmer]\n",
    "    \n",
    "    SnvCausesDramaticIncrease=False\n",
    "    DramaticIncreaseFoldChanges=[]\n",
    "    # If low aff ets, see which snvs cause dramatic affinity increase\n",
    "    for i in range(len(kmer)):\n",
    "\n",
    "        # Create all possible snvs\n",
    "        kmerDegenerate=kmer[:i]+'N'+kmer[i+1:]\n",
    "        core=kmerDegenerate[2:6]\n",
    "        \n",
    "        if core not in etsTfCores: continue \n",
    "            \n",
    "        allSnvs=jsd.GenerateAllPossibleSequences(kmerDegenerate)\n",
    "        for snv in allSnvs:\n",
    "\n",
    "            snvAff=Ets2RoundedAff[snv]\n",
    "            foldChange=snvAff/aff\n",
    "\n",
    "            # If affinity jump is sufficient, save \n",
    "            if foldChange>=foldChangeCutoff:\n",
    "                SnvCausesDramaticIncrease=True\n",
    "                DramaticIncreaseFoldChanges.append(foldChange)\n",
    "    \n",
    "    if DramaticIncreaseFoldChanges!=[]:\n",
    "        return SnvCausesDramaticIncrease,max(DramaticIncreaseFoldChanges)\n",
    "    else:\n",
    "        return SnvCausesDramaticIncrease,False\n",
    "    \n",
    "checkIfSnvCausesDramaticIncrease('AAGGAATT',Ets2RoundedAff,3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rural-stuart",
   "metadata": {},
   "source": [
    "# Option 2: Violin + pie chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "velvet-medication",
   "metadata": {},
   "outputs": [],
   "source": [
    "donut=True\n",
    "pop=True\n",
    "inner=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "continued-bulgaria",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants for barchart\n",
    "FoldChangeBin2Name={\n",
    " '[0.0, 3.0)': 'Decrease Affinity\\n or No Change',\n",
    " '[3.0, 4.0)': '3x Increase',\n",
    " '[4.0, 5.0)': '4x Increase',\n",
    " '[5.0, 6.0)': '5x Increase',\n",
    " '[6.0, 7.0)': '6x Increase',\n",
    " '[7.0, 8.0)': '7x Increase',\n",
    " '[8.0, inf)': '8x Increase',\n",
    "}\n",
    "Group2ColorAlpha={\n",
    "'3x Increase'                     :('dodgerblue',.2),\n",
    "'4x Increase'                     :('dodgerblue',.5),\n",
    "'5x Increase'                     :('dodgerblue',.7),\n",
    "'6x Increase'                     :('dodgerblue',.8),\n",
    "'7x Increase'                     :('dodgerblue',.9),\n",
    "'8x Increase'                     :('dodgerblue',1),\n",
    "'Decrease Affinity\\n or No Change':('grey',.7),\n",
    "# 'Minor Increase (1-2x)'           :('grey',.9) \n",
    "}\n",
    "\n",
    "groups=[\n",
    "'3x Increase',                     \n",
    "'4x Increase',                     \n",
    "'5x Increase',                     \n",
    "'6x Increase',                     \n",
    "'7x Increase',                     \n",
    "'8x Increase']\n",
    "#         'Decrease Affinity\\n or No Change',]\n",
    "\n",
    "def my_autopct(pct):\n",
    "    return (f'{int(pct)}%') if pct > 5 else ''\n",
    "\n",
    "ets=set(['GGAA',jsd.revcomp('GGAA'),'GGAT',jsd.revcomp('GGAT')])\n",
    "foldChangeCutoff=3\n",
    "\n",
    "beta=False\n",
    "maxBetaLines=1000\n",
    "\n",
    "allData=[]\n",
    "allDataLabels=[]\n",
    "allAverageEtsAffs=[]\n",
    "\n",
    "if beta==True: print('beta!!!   '*300,'\\n\\n\\n')\n",
    "for genome in ['Ci19','hg38']:\n",
    "\n",
    "    print('---------------------------'*3)\n",
    "    print(genome)\n",
    "    print()\n",
    "    \n",
    "    # Set up datasets\n",
    "    Name2KmerSeqAff2Count={}\n",
    "    Name2FoldChanges     ={}\n",
    "    \n",
    "    Name2NPutEns={}\n",
    "    Name2NEtsInPutEns={}\n",
    "    \n",
    "    Name2NPutEnsWithDramaticAffIncrease={}\n",
    "    Name2NEsInPutEnsWithDramaticAffIncrease={}\n",
    "    \n",
    "    #####################################################################################\n",
    "    # Analyze all datasets\n",
    "    #####################################################################################\n",
    "    \n",
    "    Name2Bed=Genome2Name2Bed[genome]\n",
    "\n",
    "    for name,bed in Name2Bed.items():\n",
    "        print('\\t',name,bed)\n",
    "        \n",
    "        # Set up this dataset\n",
    "        Name2KmerSeqAff2Count[name]={}\n",
    "        \n",
    "        Name2NPutEns[name]=0\n",
    "        Name2NEtsInPutEns[name]=0\n",
    "        \n",
    "        Name2NPutEnsWithDramaticAffIncrease[name]=0\n",
    "        Name2NEsInPutEnsWithDramaticAffIncrease[name]=0\n",
    "        \n",
    "        # Read all lines\n",
    "        MaxFoldChangeInPutEn=[]\n",
    "        \n",
    "        # Keep track of length distribution of regions\n",
    "        regionLengthDist=[]\n",
    "        \n",
    "        # Begin reading file\n",
    "        betaLineCount=0\n",
    "        for row in js.read_tsv(bed,pc=False,header=False):\n",
    "            betaLineCount+=1\n",
    "            if betaLineCount>maxBetaLines and beta==True: break\n",
    "                \n",
    "            chrom,start,end=row[:3]\n",
    "            start0=int(start)\n",
    "            end0=int(end)\n",
    "\n",
    "            regionLengthDist.append(end0-start0)\n",
    "            \n",
    "            Name2NPutEns[name]+=1\n",
    "            thisPutEnHasDramaticAffIncrease=False\n",
    "            maxEtsFoldChangedObserved=0\n",
    "            \n",
    "            \n",
    "            for kmer8 in jsd.get_kmers(G2Chr2Seq[genome][chrom][start0:end0],8):\n",
    "                \n",
    "                thisEtsHasDramaticAffIncrease=False\n",
    "\n",
    "                kmer8=kmer8.upper()\n",
    "                if 'N' in kmer8: continue\n",
    "\n",
    "                core=kmer8[2:6]\n",
    "\n",
    "                if core in ets:\n",
    "                    Name2NEtsInPutEns[name]+=1\n",
    "                    \n",
    "                    aff=Ets2RoundedAff[kmer8]\n",
    "\n",
    "                    if (kmer8,aff) not in Name2KmerSeqAff2Count[name]:\n",
    "                        Name2KmerSeqAff2Count[name][(kmer8,aff)]=0\n",
    "                    Name2KmerSeqAff2Count[name][(kmer8,aff)]+=1\n",
    "                    \n",
    "                    SnvCausesDramaticIncrease,maxFoldChange=checkIfSnvCausesDramaticIncrease(kmer8,Ets2RoundedAff,foldChangeCutoff)\n",
    "                    if SnvCausesDramaticIncrease:\n",
    "                        if maxFoldChange>maxEtsFoldChangedObserved: maxEtsFoldChangedObserved=maxFoldChange\n",
    "                        thisEtsHasDramaticAffIncrease  =True\n",
    "                        thisPutEnHasDramaticAffIncrease=True\n",
    "                        \n",
    "                        \n",
    "                if thisEtsHasDramaticAffIncrease:\n",
    "                    Name2NEsInPutEnsWithDramaticAffIncrease[name]+=1\n",
    "                    \n",
    "            if thisPutEnHasDramaticAffIncrease:\n",
    "                Name2NPutEnsWithDramaticAffIncrease[name]+=1\n",
    "                MaxFoldChangeInPutEn.append(maxEtsFoldChangedObserved)\n",
    "            else:\n",
    "                MaxFoldChangeInPutEn.append(0)\n",
    "            \n",
    "\n",
    "        nPutTotal    = Name2NPutEns[name]\n",
    "        nPutIncrease = Name2NPutEnsWithDramaticAffIncrease[name]\n",
    "        \n",
    "        print(name)\n",
    "        print(f'\\tN Put Enhancers Total                             = {nPutTotal:,}')\n",
    "        print(f'\\tN Put Enhancers With Dramatic Aff Increase        = {nPutIncrease:,} ({js.percent(nPutIncrease/nPutTotal)})')\n",
    "        print()\n",
    "\n",
    "        nEtsInPutTotal   =Name2NEtsInPutEns[name]\n",
    "        nEtsInPutIncrease=Name2NEsInPutEnsWithDramaticAffIncrease[name]\n",
    "        print(f'\\tN Ets in Put Enhancers Total                      = {nEtsInPutTotal:,}')\n",
    "        print(f'\\tN Ets in Put Enhancers With Dramatic Aff Increase = {nEtsInPutIncrease:,} ({js.percent(nEtsInPutIncrease/nEtsInPutTotal)})')\n",
    "        print()\n",
    "        \n",
    "\n",
    "\n",
    "        ################################################################\n",
    "        # Determine all possible SNVs which create dramatic change\n",
    "        ################################################################\n",
    "\n",
    "        namePlusSnv='+SNV '+name\n",
    "        Name2KmerSeqAff2Count[namePlusSnv]={}\n",
    "        Name2FoldChanges[namePlusSnv]=[]\n",
    "\n",
    "        for (kmer,aff),count in Name2KmerSeqAff2Count[name].items():\n",
    "\n",
    "            # If low aff ets, see which snvs cause dramatic affinity increase\n",
    "            for i in range(len(kmer)):\n",
    "\n",
    "                # Create all possible snvs\n",
    "                kmerDegenerate=kmer[:i]+'N'+kmer[i+1:]\n",
    "\n",
    "                # Only keep snv if snv is in flank\n",
    "                core=kmerDegenerate[2:6]\n",
    "                if 'N' in core: continue\n",
    "\n",
    "                allSnvs=jsd.GenerateAllPossibleSequences(kmerDegenerate)\n",
    "                for snv in allSnvs:\n",
    "\n",
    "                    snvAff=Ets2RoundedAff[snv]\n",
    "                    foldChange=snvAff/aff\n",
    "                    Name2FoldChanges[namePlusSnv].append(snvAff/aff)\n",
    "\n",
    "                    # If affinity jump is sufficient, save \n",
    "                    if foldChange>=foldChangeCutoff:\n",
    "\n",
    "                        if (snv,snvAff) not in Name2KmerSeqAff2Count[namePlusSnv]:\n",
    "                            Name2KmerSeqAff2Count[namePlusSnv][(snv,snvAff)]=0\n",
    "                        Name2KmerSeqAff2Count[namePlusSnv][(snv,snvAff)]+=count\n",
    "\n",
    "        ################################################################\n",
    "        # Violiin plot\n",
    "        ################################################################\n",
    "\n",
    "            \n",
    "        # Data\n",
    "        Name2AffArray={}\n",
    "        for dname in Name2KmerSeqAff2Count:\n",
    "            Name2AffArray[dname]=[]\n",
    "\n",
    "            for (kmer,aff),count in Name2KmerSeqAff2Count[dname].items():\n",
    "                Name2AffArray[dname]+=[aff]*count\n",
    "\n",
    "        snvLabel='+SNVs Opt Aff ≥3-fold\\n'\n",
    "        data=[Name2AffArray[name],Name2AffArray[namePlusSnv]]\n",
    "        allData+=data\n",
    "        \n",
    "        avgOpenAff = round(np.median(Name2AffArray[name]),2)\n",
    "        avgSnvAff  = round(np.median(Name2AffArray[namePlusSnv]),2)\n",
    "\n",
    "        print(f'Average Putative Dev Enhancer ETS     = {avgOpenAff}')\n",
    "        print(f'Average Putative Dev Enhancer ETS-SNV = {avgSnvAff}')\n",
    "        \n",
    "        for text in [True,False]:\n",
    "\n",
    "            # Figure\n",
    "            fig,ax=plt.subplots(1,2,figsize=(10,5),dpi=300,facecolor='white',gridspec_kw={'width_ratios':[1.7,1], 'height_ratios':[1]})\n",
    "\n",
    "            ############################################\n",
    "            # Violin\n",
    "            ############################################\n",
    "            \n",
    "            ax1=ax[0]\n",
    "            x=sns.violinplot(data=data,ax=ax1,scale='count',palette='Blues',cut=0)\n",
    "            \n",
    "            ############################################\n",
    "            # Plot dots of disease kmers\n",
    "            ############################################\n",
    "            print(Genome2Name2DiseaseKmers[genome][name])\n",
    "            diseaseKmers,xAdjustCustom=Genome2Name2DiseaseKmers[genome][name]\n",
    "            \n",
    "            xAdjustScale=.025\n",
    "            xcustom=[1.45+i*xAdjustScale for i in xAdjustCustom]\n",
    "            \n",
    "            diseaseKmerAffs=[Ets2RoundedAff[kmer] for kmer in diseaseKmers]\n",
    "            x=ax1.scatter(x=xcustom,y=diseaseKmerAffs,color='white',marker='o',edgecolor='dimgrey',s=28)\n",
    "            \n",
    "            # AES\n",
    "            if text:\n",
    "                x=ax1.set_ylabel('ETS Affinity\\n',size=15)\n",
    "                x=ax1.set_title(genome+'\\n\\n\\n\\n\\n')\n",
    "                \n",
    "            ax1.spines[\"right\"].set_visible(False)\n",
    "            ax1.spines[\"top\"].set_visible(False)\n",
    "\n",
    "            # dashed line at average affinity\n",
    "            ax1.hlines(y=avgOpenAff, xmin=0, xmax=1.15, linewidth=2, color='dodgerblue',alpha=.3,ls='--')\n",
    "            ax1.hlines(y=avgSnvAff, xmin=1, xmax=1.15, linewidth=2,  color='dodgerblue',alpha=1,ls='--')\n",
    "            \n",
    "            allAverageEtsAffs+=[avgOpenAff,avgSnvAff]\n",
    "            \n",
    "            if text:\n",
    "                ax1.text(1.15,avgOpenAff,avgOpenAff,va='center',color='dodgerblue',alpha=.5)\n",
    "                ax1.text(1.15,avgSnvAff,avgSnvAff,va='center',color='dodgerblue',alpha=1)\n",
    "\n",
    "            # xticklabels\n",
    "            if text:\n",
    "                xlabels=[Name2Xlabel[name],snvLabel+Name2Xlabel[name]]\n",
    "                allDataLabels+=xlabels\n",
    "                x=ax1.set_xticklabels(xlabels)\n",
    "\n",
    "            # Plot how many regions were able to be optimized\n",
    "            percentRegionsOptimized=f'{round(100*nPutIncrease/nPutTotal,1)}%'\n",
    "            if text:\n",
    "                ax1.text(1,1.15,f'\\n{percentRegionsOptimized} ({nPutIncrease:,} of {nPutTotal:,}) \\nregions contain ≥1 optimizable Ets',ha='center',color='dodgerblue',va='bottom')\n",
    "\n",
    "            #################################################################\n",
    "            # Pie Chart\n",
    "            #################################################################\n",
    "\n",
    "            # Set xlim to allow barplot to be included\n",
    "            if text: ax1.set_xlim(-.5,2)\n",
    "\n",
    "            data_pie=pd.Series(MaxFoldChangeInPutEn)\n",
    "            data_pie=pd.cut(data_pie,bins=[0,3,4,5,6,7,8,np.inf],right=False).value_counts()\n",
    "            FoldChangeGroup2Count={str(b):c for b,c in zip(data_pie.index,data_pie.tolist())}\n",
    "\n",
    "            FoldChangeGroup2Count={FoldChangeBin2Name[b]:c for b,c in FoldChangeGroup2Count.items()}\n",
    "\n",
    "            labels=[g for g in groups]\n",
    "            counts=[FoldChangeGroup2Count[g] for g in groups]\n",
    "            alphas=[Group2ColorAlpha[g][1] for g in groups]\n",
    "            colors=[(72/255, 195/255, 68/255, a) for a in alphas]\n",
    "            colors=[(0/255, 128/255, 255/255, a) for a in alphas]\n",
    "            print(alphas)\n",
    "\n",
    "            if pop: explode = [.025 for g in groups]\n",
    "            else:   explode = [0 for g in groups]\n",
    "\n",
    "            # pie            \n",
    "            ax2=ax[1]\n",
    "            if text: ax2.pie(counts, explode=explode, labels=[l[:2] for l in labels],shadow=False, startangle=90,colors=colors, autopct=lambda pct: f\"{pct:.1f}%\" if pct>7 else '' )\n",
    "            else:    ax2.pie(counts, explode=explode, labels=['' for l in labels],shadow=False, startangle=90,colors=colors)\n",
    "            if text:\n",
    "                ax2.set_title('Percent of Putative Enhancers\\nGrouped by Fold Change')\n",
    "\n",
    "            # remove all text\n",
    "            if not text:\n",
    "                for a in [ax1,ax2]:\n",
    "                    a.set_yticklabels([])\n",
    "                    a.set_xticklabels([])\n",
    "\n",
    "\n",
    "\n",
    "            plt.show()\n",
    "\n",
    "            figname=name.replace('\\n',' ').replace('(','').replace(')','').replace('.','')\n",
    "            fig.savefig(f\"{figname}.text={text}.violin.svg\", format=\"svg\",dpi=300,bbox_inches='tight')\n",
    "            fig.savefig(f\"{figname}.text={text}.violin.png\", format=\"png\",dpi=300,bbox_inches='tight')\n",
    "\n",
    "\n",
    "        # hist of region lengths\n",
    "        plt.hist(regionLengthDist,bins=200)\n",
    "        plt.title(name)\n",
    "        plt.xlabel('Region Length')\n",
    "        plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff885bfd-4b50-40a4-98ef-7cbdd4b88291",
   "metadata": {},
   "source": [
    "# Plot on same x axis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d535eab7-09d0-4fcc-a1a0-d5b8d081333d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recall you have the following data\n",
    "# allData=[]\n",
    "# allDataLabels=[]\n",
    "# allAverageEtsAffs=[]\n",
    "\n",
    "for text in [True,False]:\n",
    "\n",
    "    fig,ax=plt.subplots(1,figsize=(11,5),dpi=200)\n",
    "\n",
    "    sns.violinplot(data=allData,ax=ax,scale='area',palette=['lightblue','royalblue']*2,cut=0)\n",
    "\n",
    "    for i,avgOpenAff in enumerate([allAverageEtsAffs[0],'NA',allAverageEtsAffs[2],'NA']):\n",
    "        if avgOpenAff=='NA': continue\n",
    "        ax.hlines(y=avgOpenAff, xmin=i+0, xmax=i+.55, linewidth=2, color='darkgrey',alpha=.3,ls='--')\n",
    "        if text: ax.text(y=avgOpenAff,x=i+.55,s=avgOpenAff,va='center',color='darkgrey')\n",
    "\n",
    "    for i,avgSnvAff in enumerate([allAverageEtsAffs[1],'NA',allAverageEtsAffs[3]]):\n",
    "        if avgSnvAff=='NA': continue\n",
    "        ax.hlines(y=avgSnvAff, xmin=i+1, xmax=i+1.15, linewidth=2,  color='grey',alpha=1,ls='--')\n",
    "        if text: ax.text(y=avgSnvAff,x=i+1.15,s=avgSnvAff,va='center',color='grey')\n",
    "\n",
    "    if text: ax.set_xticklabels(allDataLabels)\n",
    "    if text: ax.set_ylabel('ETS Affinity')\n",
    "    \n",
    "    else:    ax.set_xticklabels([])\n",
    "    if not text: ax.set_yticklabels([])\n",
    "    \n",
    "    plt.show()\n",
    "    \n",
    "    fig.savefig(f\"merged-xaxis.text={text}.violin.png\", format=\"png\",dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa522e1e-581c-42d8-bcea-06027f307de4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
