{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ddd74fa3",
   "metadata": {},
   "source": [
    "# Looking at disease associated SNVs wrt ETS affinity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c39895d-5693-446d-803c-d1357d349b84",
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import time\n",
    "import pickle\n",
    "import pybedtools\n",
    "import js\n",
    "import os\n",
    "\n",
    "#########################################################################################\n",
    "# These methods can be found on my gitub\n",
    "# see https://github.com/jsolvason/js/tree/6623ca2c3ee871eca8defbac5b85fcb4036eb7a4\n",
    "#    for methods used here at time of publication\n",
    "\n",
    "import jsGenome as jsg\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f6a810d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def setprint(s,n):\n",
    "    x=0\n",
    "    for i in s:\n",
    "        x+=1\n",
    "        print(i)\n",
    "        if x>n: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08e60e4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################################################\n",
    "# Download data form wei 2010, \n",
    "# http://the_brain.bwh.harvard.edu/uniprobe/detailsDef.php?id=414\n",
    "# Mouse ETS1 pbm data\n",
    "\n",
    "# Ets1\n",
    "ref  =f'/Users/joe/code/ref/binding_affinity/ets/parsed_Ets1_8mers.txt'              \n",
    "Seq2EtsAff  = {line.split('\\t')[0]:float(line.split('\\t')[1]) for line in open(ref,'r').readlines()}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee8cc85e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "hg2chr2seq={}\n",
    "\n",
    "# Load hg38  genome\n",
    "# from Bio import SeqIO\n",
    "file_genome=f'/Users/joe/code/ref/genomes/human/hg38/hg38.fa.chr2seq.pydict.pickle'\n",
    "hg2chr2seq['hg38'] = jsg.pklLoadGenome(file_genome)\n",
    "\n",
    "# Load hg19 genome\n",
    "file_genome=f'/Users/joe/code/ref/genomes/human/hg19/hg19.fa.chr2seq.pydict.pickle'\n",
    "hg2chr2seq['hg19'] = jsg.pklLoadGenome(file_genome)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c40671e9",
   "metadata": {},
   "source": [
    "# Read Datasests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c971113",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAll8mers(seq,p,ref,mut):\n",
    "    \n",
    "    seq=seq[p-7:p+8]    \n",
    "    seq=seq[:7]+ref+seq[7+1:]\n",
    "    \n",
    "    # print(seq[p])\n",
    "    for i,end in enumerate(range(7,len(seq))):\n",
    "        start=end-7\n",
    "        last=end+1\n",
    "        kmer8=seq[start:last]\n",
    "        if i==0: kmer8mut=kmer8[:-1]+mut\n",
    "        else:    kmer8mut=kmer8[:-i-1]+mut+kmer8[-i:]\n",
    "            \n",
    "        yield kmer8,kmer8mut\n",
    "                    #.  012345678901234567890\n",
    "            \n",
    "print(list(getAll8mers('NNNNNNGGAANNNNNNNNN',10,'G','X')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3aeb301",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ets(kmer4):\n",
    "    if kmer4 in ['GGAA','GGAT','TTCC','ATCC']: return True\n",
    "    else:                                      return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b40be900",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up logic for creating/deleting ets sites\n",
    "BeforeAfter2EtsLogic={\n",
    "    (0,0):'Ets Never',\n",
    "    (1,1):'Ets Always',\n",
    "    (0,1):'Ets Created',\n",
    "    (1,0):'Ets Deleted'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74b7fbee",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "###################################################################################\n",
    "# Set up input files\n",
    "###################################################################################\n",
    "\n",
    "# All data must be hg38 or lifted to hg38\n",
    "# gwas datasets already filtered for pval ≤ 5e-8\n",
    "Data2Fn={\n",
    "    'gwas_uk_pulse_rate'           :f'./0-data/0-snvs/ukbiobank-gwas-hg19/hg38/pvalues-≤5e-8/continuous-102-both_sexes-irnt-hg38-pval≤5e-08.tsv',\n",
    "    'gwas_uk_qrs_duration'         :f'./0-data/0-snvs/ukbiobank-gwas-hg19/hg38/pvalues-≤5e-8/continuous-12340-both_sexes-irnt-hg38-pval≤5e-08.tsv',\n",
    "    'gwas_uk_pulse_pressure'       :f'./0-data/0-snvs/ukbiobank-gwas-hg19/hg38/pvalues-≤5e-8/continuous-PP-both_sexes-combined_medadj_irnt-hg38-pval≤5e-08.tsv',\n",
    "    'gwas_uk_afib'                 :f'./0-data/0-snvs/ukbiobank-gwas-hg19/hg38/pvalues-≤5e-8/icd10-I48-both_sexes-hg38-pval≤5e-08.tsv',\n",
    "    'gwas_uk_myocardial_infarction':f'./0-data/0-snvs/ukbiobank-gwas-hg19/hg38/pvalues-≤5e-8/phecode-411-hg38-pval≤5e-08.tsv',\n",
    "}\n",
    "\n",
    "Data2Tuple2Info={} # tuple=(chrom,0idx,genome,Ref,Alt,infoTuple)\n",
    "Data2SampleSize={}\n",
    "Data2Allele2Count={}\n",
    "    \n",
    "###################################################################################\n",
    "# Read inputs\n",
    "###################################################################################\n",
    "\n",
    "sanityCheck=False\n",
    "lineCount_sanityCheck=100000\n",
    "for dataName,fn in Data2Fn.items():\n",
    "    \n",
    "    lc=0\n",
    "    \n",
    "    print(dataName)\n",
    "    \n",
    "    # Create datasets\n",
    "    Data2Tuple2Info[dataName]={}\n",
    "    Data2SampleSize[dataName]={'all':0,'skipped':0,'chosen':0,'duplicate':0}\n",
    "    Data2Allele2Count[dataName]={'all':0,'ref':0,'alt':0,'neither':0}\n",
    "    \n",
    "    # Read file\n",
    "    for row in js.read_tsv(fn,pc=False,header=True):\n",
    "        lc+=1\n",
    "        \n",
    "        Data2SampleSize[dataName]['all']+=1\n",
    "        \n",
    "        # Parse row\n",
    "        chrom,idx1,ref,alt,info = row\n",
    "        idx0=int(idx1)-1\n",
    "        if 'chr' not in chrom: chrom=f'chr{chrom}'\n",
    "        genome='hg38'\n",
    "\n",
    "        # Ensure data is snp, no indels\n",
    "        if len(ref)!=1 or len(alt)!=1: \n",
    "            Data2SampleSize[dataName]['skipped']+=1\n",
    "            continue\n",
    "\n",
    "        # Tabulate what proportion of nt are ref or alt in genome\n",
    "        snvNtInGenome=hg2chr2seq[genome][chrom][idx0]\n",
    "        Data2Allele2Count[dataName]['all']+=1\n",
    "        if ref==snvNtInGenome:    Data2Allele2Count[dataName]['ref']+=1 \n",
    "        elif alt==snvNtInGenome:  Data2Allele2Count[dataName]['alt']+=1\n",
    "        else:                     Data2Allele2Count[dataName]['neither']+=1 \n",
    "        if sanityCheck==True and lc>lineCount_sanityCheck: break\n",
    "                \n",
    "        # Ensure snp changes ets affinity\n",
    "        etsChanged=False\n",
    "        for kmer8,kmer8mut in getAll8mers(hg2chr2seq[genome][chrom],idx0,ref,alt):\n",
    "            core   =kmer8[2:6]\n",
    "            coremut=kmer8mut[2:6]\n",
    "            \n",
    "            ba=(ets(core),ets(coremut)) # before after logic\n",
    "            etsLogic=BeforeAfter2EtsLogic[ba]\n",
    "            \n",
    "            # If ets-changing snp found, continue\n",
    "            if etsLogic=='Ets Always': \n",
    "                etsChanged=True\n",
    "                break            \n",
    "\n",
    "        if etsChanged:\n",
    "\n",
    "            # Add data if not duplicate\n",
    "            dataTuple=(chrom,idx0,genome,ref,alt)\n",
    "            if dataTuple not in Data2Tuple2Info[dataName]:\n",
    "                Data2SampleSize[dataName]['chosen']+=1\n",
    "                Data2Tuple2Info[dataName][dataTuple]=set()\n",
    "                Data2Tuple2Info[dataName][dataTuple].add(info)\n",
    "\n",
    "            else:\n",
    "                Data2SampleSize[dataName]['duplicate']+=1\n",
    "                Data2Tuple2Info[dataName][dataTuple].add(info)\n",
    "\n",
    "\n",
    "    print(dataName, f'{len(Data2Tuple2Info[dataName]):,} variants added')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4be821d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#####################################################\n",
    "# Sanity check that REF or ALT is at expected loc\n",
    "#####################################################\n",
    "\n",
    "for data in Data2Allele2Count:\n",
    "    print('--------'*5,'\\n\\n')\n",
    "    print(data)\n",
    "    print()\n",
    "    \n",
    "    n=Data2Allele2Count[data]['all']\n",
    "    \n",
    "    if n==0: \n",
    "        print('No variants in dataset, skipping\\n')\n",
    "        continue\n",
    "        \n",
    "    nRef=Data2Allele2Count[data]['ref']\n",
    "    nAlt=Data2Allele2Count[data]['alt']\n",
    "    nNone=Data2Allele2Count[data]['neither']\n",
    "    \n",
    "    refRate  = nRef/n\n",
    "    altRate  = nAlt/n\n",
    "    noneRate = nNone/n\n",
    "    \n",
    "\n",
    "    print(f'\\tRef Rate     = {js.percent(refRate                 , 2)}')\n",
    "    print(f'\\tAlt Rate     = {js.percent(altRate                 , 2)}')\n",
    "    print(f'\\tNeither Rate = {js.percent(noneRate                , 2)}')\n",
    "    print(f'\\tTotal %      = {js.percent(refRate+altRate+noneRate, 2)}')\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5004bdb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for d in Data2SampleSize:\n",
    "    nAll=Data2SampleSize  [d]['all']\n",
    "    nSkip=Data2SampleSize [d]['skipped']\n",
    "    nChose=Data2SampleSize[d]['chosen']\n",
    "    \n",
    "    print('-----------'*5)\n",
    "    print(d)\n",
    "    if nAll>0:\n",
    "        print(f'\\tN        = {nAll:,}')\n",
    "        print(f'\\tChosen   = {nChose:,}')    \n",
    "        print(f'\\t% Chosen = {js.percent(nChose/nAll,2)}')    \n",
    "    else:\n",
    "        print('None found, skipping')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08013b38",
   "metadata": {},
   "outputs": [],
   "source": [
    "for d in Data2Tuple2Info:\n",
    "    \n",
    "    print('-----------'*5)\n",
    "    print(d)\n",
    "    js.dprint(Data2Tuple2Info[d],2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e92d2dc",
   "metadata": {},
   "source": [
    "# Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bd2d942",
   "metadata": {},
   "outputs": [],
   "source": [
    "################################################################\n",
    "# Set up datatable\n",
    "################################################################\n",
    "\n",
    "columns=['dataset'       ,\n",
    "         'chrom'         ,\n",
    "         'idx0'  ,\n",
    "         'idx1',\n",
    "         'ref'     ,\n",
    "         'alt'    ,\n",
    "         'ref8mer',\n",
    "         'risk8mer',\n",
    "#          'effect'        ,\n",
    "#          'creEts'        ,\n",
    "#          'creEtsAff'     ,\n",
    "#          'delEts'        ,\n",
    "#          'delEtsAff'     ,\n",
    "#          'changedEts'    ,\n",
    "         'refAff',\n",
    "         'riskAff'     ,\n",
    "         'change-in-affinity',\n",
    "         'info-list',\n",
    "#          'genome'\n",
    "        ]\n",
    "c2v={c:[] for c in columns}\n",
    "\n",
    "\n",
    "################################################################\n",
    "# Set up sanity check for indexing snp location\n",
    "################################################################\n",
    "\n",
    "Data2Allele2Count={}\n",
    "\n",
    "################################################################\n",
    "# For each dataset\n",
    "################################################################\n",
    "\n",
    "for data in Data2Tuple2Info:\n",
    "    print(data)\n",
    "        \n",
    "    ############################################################\n",
    "    # For each snp\n",
    "    ############################################################\n",
    "    \n",
    "    # (chrom,idx0,ref,alt,genome,infoList)\n",
    "    for (c,p,genome,ref,alt),infoList in Data2Tuple2Info[data].items():\n",
    "        \n",
    "        ############################################################\n",
    "        # Consider each 8mer of the snp\n",
    "        ############################################################\n",
    "        for kmer8,kmer8mut in getAll8mers(hg2chr2seq[genome][c],p,ref,alt):\n",
    "            core   =kmer8[2:6]\n",
    "            coremut=kmer8mut[2:6]\n",
    "            \n",
    "            ba=(ets(core),ets(coremut)) # before after logic\n",
    "            etsLogic=BeforeAfter2EtsLogic[ba]\n",
    "            \n",
    "            if etsLogic=='Ets Never':\n",
    "#                 c2v['dataset'       ].append(data)\n",
    "#                 c2v['chrom'         ].append(c)\n",
    "#                 c2v['pos'           ].append(p)\n",
    "#                 c2v['ref'     ].append(ref)\n",
    "#                 c2v['alt'    ].append(alt)\n",
    "#                 c2v['effect'        ].append(etsLogic)\n",
    "#                 c2v['ref8mer'       ].append(kmer8)\n",
    "#                 c2v['risk8mer'      ].append(kmer8mut)\n",
    "#                 c2v['creEts'        ].append(False)\n",
    "#                 c2v['creEtsAff'     ].append(False)\n",
    "#                 c2v['delEts'        ].append(False)\n",
    "#                 c2v['delEtsAff'     ].append(False)\n",
    "#                 c2v['changesAff'    ].append(False)\n",
    "                pass\n",
    "                \n",
    "            elif etsLogic=='Ets Always':\n",
    "                aff   =Seq2EtsAff[kmer8]\n",
    "                if ('?' not in kmer8mut) and ('*' not in kmer8mut):  \n",
    "                    affmut=Seq2EtsAff[kmer8mut]\n",
    "                    delta=affmut-aff\n",
    "                else:                    \n",
    "                    affmut=False\n",
    "                    delta=False\n",
    "                c2v['dataset'             ].append(data)\n",
    "                c2v['chrom'               ].append(c)\n",
    "                c2v['idx0'                ].append(p)\n",
    "                c2v['idx1'                ].append(p+1)\n",
    "                c2v['ref'                 ].append(ref)\n",
    "                c2v['alt'                 ].append(alt)\n",
    "#                 c2v['effect'              ].append(etsLogic)\n",
    "                c2v['ref8mer'             ].append(kmer8)\n",
    "                c2v['risk8mer'            ].append(kmer8mut)\n",
    "#                 c2v['creEts'              ].append(False)\n",
    "#                 c2v['creEtsAff'           ].append(False)\n",
    "#                 c2v['delEts'              ].append(False)\n",
    "#                 c2v['delEtsAff'           ].append(False)\n",
    "#                 c2v['changedEts'          ].append(1)\n",
    "                c2v['refAff'              ].append(aff)\n",
    "                c2v['riskAff'             ].append(affmut)\n",
    "                c2v['change-in-affinity'  ].append(delta)\n",
    "                c2v['info-list'           ].append(infoList)\n",
    "#                 c2v['genome'              ].append(genome)\n",
    "                \n",
    "            elif etsLogic=='Ets Created':\n",
    "#                 affmut=Seq2EtsAff[kmer8mut]\n",
    "#                 c2v['dataset'             ].append(data)\n",
    "#                 c2v['chrom'               ].append(c)\n",
    "#                 c2v['idx0'        ].append(p)\n",
    "#                 c2v['ref'           ].append(ref)\n",
    "#                 c2v['alt'          ].append(alt)\n",
    "#                 c2v['effect'              ].append(etsLogic)\n",
    "#                 c2v['ref8mer'             ].append(kmer8)\n",
    "#                 c2v['risk8mer'            ].append(kmer8mut)\n",
    "#                 c2v['creEts'              ].append(1)\n",
    "#                 c2v['creEtsAff'           ].append(affmut)\n",
    "#                 c2v['delEts'              ].append(False)\n",
    "#                 c2v['delEtsAff'           ].append(False)\n",
    "#                 c2v['changedEts'          ].append(False)\n",
    "#                 c2v['refAff'              ].append(False)\n",
    "#                 c2v['riskAff'             ].append(False)\n",
    "#                 c2v['change-in-affinity'      ].append(False)\n",
    "                pass\n",
    "                \n",
    "            elif etsLogic=='Ets Deleted':\n",
    "#                 aff=Seq2EtsAff[kmer8]\n",
    "#                 c2v['dataset'             ].append(data)\n",
    "#                 c2v['chrom'               ].append(c)\n",
    "#                 c2v['idx0'        ].append(p)\n",
    "#                 c2v['ref'           ].append(ref)\n",
    "#                 c2v['alt'          ].append(alt)\n",
    "#                 c2v['effect'              ].append(etsLogic)\n",
    "#                 c2v['ref8mer'             ].append(kmer8)\n",
    "#                 c2v['risk8mer'            ].append(kmer8mut)\n",
    "#                 c2v['creEts'              ].append(False)\n",
    "#                 c2v['creEtsAff'           ].append(False)\n",
    "#                 c2v['delEts'              ].append(1)\n",
    "#                 c2v['delEtsAff'           ].append(aff)\n",
    "#                 c2v['changedEts'          ].append(False)\n",
    "#                 c2v['refAff'              ].append(False)\n",
    "#                 c2v['riskAff'             ].append(False)\n",
    "#                 c2v['change-in-affinity'      ].append(False)\n",
    "                pass\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c45830ae",
   "metadata": {},
   "source": [
    "# Create DF\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6810963e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pd.DataFrame(c2v)\n",
    "df['chrom-pos1-ref-alt-tuple']=df.apply(lambda row: (row['chrom'],row['idx1'],row['ref'],row['alt']),axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2e02feb",
   "metadata": {},
   "source": [
    "# Generate Columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d72c533",
   "metadata": {},
   "source": [
    "## REF - Load reference data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0784a7c9",
   "metadata": {},
   "source": [
    "### REF - Func to print % True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f21c4c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "def percentTrue(l,col):\n",
    "    boolList=[]\n",
    "    for i in l:\n",
    "        if i!=False: boolList.append(True)\n",
    "        else:        boolList.append(False)\n",
    "    n=sum(boolList)\n",
    "    p=js.percent(sum(boolList) / len(boolList),2)\n",
    "    print(f'{n:,} True ({p}) in {col}')\n",
    "    \n",
    "percentTrue([False,1,1,1],'test')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a34451a-a4ae-435a-a78f-46b45c31f29d",
   "metadata": {},
   "source": [
    "### REF - GeneID to Name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7ff1d5b-3dc1-44e4-9f91-539a3ebdebc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./0-data/0-snvs/dantonio-eqtl-hg19/hg19/table_s2_SnpChromPos2Genes.pydict.pickle',\"rb\") as f:\n",
    "    EqtlChromPos2Genes=pickle.load(f)\n",
    "\n",
    "# with open('./0-data/0-snvs/dantonio-eqtl-hg19/hg19/table_s2_GeneId2Name.pydict.pickle','rb') as f:\n",
    "#     GeneId2Name=pickle.load(f)\n",
    "\n",
    "with open('/Users/joe/code/ref/genomes/human/hg19/other/gencode.v34lift37.GeneId2Name.pydict.pickle','rb') as f:\n",
    "    GeneId2Name=pickle.load(f)\n",
    "    \n",
    "js.dprint(GeneId2Name,0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "196f1ae1-ee01-4004-b372-cf0f2e817966",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def getGeneName(geneId):\n",
    "#     geneIdBasename=geneId.split('.')[0]\n",
    "#     if geneIdBasename in GeneId2Name:\n",
    "#         geneName=GeneId2Name[geneIdBasename]\n",
    "#         # Don't give nonsense names - currently set up for hg38.ncbiRefSeq.gtf\n",
    "#         if 'LOC' not in geneName:\n",
    "#             return geneName\n",
    "#         else:\n",
    "#             return 'NoName'\n",
    "#     else:\n",
    "#         return 'NoName'\n",
    "    \n",
    "# getGeneName('ENSG00000128309.13_6'),getGeneName('ENSG000001')\n",
    "\n",
    "def getGeneName(geneId):\n",
    "    if geneId in GeneId2Name:\n",
    "        geneName=GeneId2Name[geneId]\n",
    "        # Don't give nonsense names - currently set up for hg38.ncbiRefSeq.gtf\n",
    "        if 'LOC' not in geneName:\n",
    "            return geneName\n",
    "        else:\n",
    "            return 'NoName'\n",
    "    else:\n",
    "        return 'NoName'\n",
    "getGeneName('ENSG00000223972.5_3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c113a3ee-a765-4072-ad77-22cb7dada530",
   "metadata": {},
   "outputs": [],
   "source": [
    "js.dprint(EqtlChromPos2Genes,0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dc50653-f4f7-4c49-9f2f-cd1ea699591e",
   "metadata": {},
   "outputs": [],
   "source": [
    "js.dprint(GeneId2Name,0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdc308d2",
   "metadata": {},
   "source": [
    "## AFF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfb566b4",
   "metadata": {},
   "source": [
    "### AFF: Delta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97570fa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare for Sorting by entries which change ets affinity by a large number\n",
    "df['change-in-affinity']=df['change-in-affinity'].apply(lambda x: x if x!=False else 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7d11d36",
   "metadata": {},
   "source": [
    "### AFF: Large Delta?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f026f31d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create column indicating if DNV creates affinity change\n",
    "foldChangeThreshold=3\n",
    "df['affinity-fold-change']=df.apply(lambda row: row['riskAff']/row['refAff'] if (type(row['refAff'])==float and type(row['riskAff'])==float) else False, axis=1)\n",
    "df['dramatic-affinity-increase']=df['affinity-fold-change'].apply(lambda foldchange: foldchange>=foldChangeThreshold)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e253aab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "col='dramatic-affinity-increase'\n",
    "f'{sum(df[col]!=False):,}',js.percent(sum(df[col]!=False) / len(df),2),f'True in {col}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4290814",
   "metadata": {},
   "outputs": [],
   "source": [
    "col='affinity-fold-change'\n",
    "f'{sum(df[col]!=False):,}',js.percent(sum(df[col]!=False) / len(df),2),f'True in {col}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "532aa183",
   "metadata": {},
   "outputs": [],
   "source": [
    "Chrom2AllChromTss={}\n",
    "for chrom,pos in hg38_ChromTss2GeneSet:\n",
    "    \n",
    "    if chrom not in Chrom2AllChromTss:\n",
    "        Chrom2AllChromTss[chrom]=set()\n",
    "    Chrom2AllChromTss[chrom].add((chrom,pos))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67642a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "js.dprint(hg38_ChromTss2GeneSet,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3698a0df",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nearbyGenes(queryChromPos,window,ChromTss2GeneSet,Chrom2AllChromTss,geneLimit=10):\n",
    "\n",
    "    qc,qp=list(queryChromPos) # chrom and pos of query\n",
    "    window=int(window)        # how far to look from tss    \n",
    "    \n",
    "    try:               qp=int(qp)\n",
    "    except ValueError: return False\n",
    "        \n",
    "    # Create datasets to fill\n",
    "    nearGenes=set()\n",
    "    nearGenesWithDistance=[]\n",
    "    \n",
    "    # Filter out all bp that are not tss\n",
    "    searchNts=[(qc,nt) for nt in range(qp-window,qp+window+1)]\n",
    "    searchNts=set(searchNts)\n",
    "    searchNts = searchNts & Chrom2AllChromTss[qc]\n",
    "        \n",
    "    # Look within search window\n",
    "    for qc,ntPos in searchNts:\n",
    "\n",
    "        # Check each possible name \n",
    "        geneList=ChromTss2GeneSet[(qc,ntPos)]\n",
    "        \n",
    "        for gene in geneList:\n",
    "\n",
    "            # Skip conditions\n",
    "            if gene[:3]=='LOC':   continue # If gene name isn't giberish => currently set up for 'hg38.ncbiRefSeq.gtf.ChrTss2GeneSet.pydict.pickle'\n",
    "            if gene in nearGenes: continue # If gene hasn't been seen before\n",
    "\n",
    "            nearGenes.add(gene)\n",
    "            distance=qp-ntPos\n",
    "            nearGenesWithDistance.append((distance,gene))\n",
    "      \n",
    "    # Convert distances to kb\n",
    "    nearGenesWithDistance=[(int((d)/1000),g) for (d,g) in nearGenesWithDistance]\n",
    "    nearGenesWithDistance=sorted(nearGenesWithDistance,key=lambda tup: abs(tup[0]),reverse=False)\n",
    "    nearGenesWithDistance=[(f'{d:,} kb',g) for (d,g) in nearGenesWithDistance]\n",
    "    \n",
    "    if geneLimit!=np.inf:\n",
    "        nearGenesWithDistance=nearGenesWithDistance[:geneLimit]\n",
    "    \n",
    "    return nearGenesWithDistance\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "get_nearbyGenes(('chr12',6534517),500e3,hg38_ChromTss2GeneSet,Chrom2AllChromTss,geneLimit=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "244af0aa",
   "metadata": {},
   "source": [
    "### NEAREST GENE: Nearest genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dbeadfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add 10 nearest genes\n",
    "nGenes=20\n",
    "window=500e3\n",
    "\n",
    "nearGeneDfRow=[]\n",
    "x=0\n",
    "n=sum((df['dramatic-affinity-increase']==True))\n",
    "start=time.time()\n",
    "\n",
    "tempSet=set()\n",
    "\n",
    "for idx in df.index:\n",
    "    chrom=df.at[idx,'chrom']\n",
    "    dnvPos=df.at[idx,'idx0']\n",
    "\n",
    "    # Determine if you will check this variant\n",
    "    worthChecking=df.at[idx,'dramatic-affinity-increase']\n",
    "    if worthChecking!=True: \n",
    "        nearGeneDfRow.append(False)\n",
    "        continue\n",
    "        \n",
    "    x+=1\n",
    "    if x%100==0: \n",
    "        end = time.time()\n",
    "        sec = end-start\n",
    "        rowsLeft=n-x\n",
    "        rowsPerSec=x/sec\n",
    "        secLeft=rowsLeft/rowsPerSec\n",
    "        minLeft=round(secLeft/60,1)\n",
    "        js.percentDone=round(100*x/n)\n",
    "        print(f'{js.percentDone}% ({minLeft} min left)',end=', ')\n",
    "        \n",
    "    nearGenesWithDistance=get_nearbyGenes((chrom,dnvPos),window,hg38_ChromTss2GeneSet,Chrom2AllChromTss,geneLimit=np.inf)\n",
    "    if nearGenesWithDistance!=[]:\n",
    "        nearGeneDfRow.append(nearGenesWithDistance)\n",
    "    else:\n",
    "        nearGeneDfRow.append(False)\n",
    "    \n",
    "df['near-genes-500kb']=nearGeneDfRow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "487e601a",
   "metadata": {},
   "outputs": [],
   "source": [
    "col='near-genes-500kb'\n",
    "percentTrue(df[col],col)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
